### PR TITLE
Handle otherName SANs in CSRs

### LINKED ISSUE #5989: Vault PKI ignores otherName field in CSR
**Describe the bug**
Clients can currently submit a CSR where various fields are specified in either the CSR blob or in the JSON payload. However, the `otherName` field is inconsistent. If there is an `otherName` value in the CSR blob, it gets ignored. Clients are relegated to only using the JSON blob.

**To Reproduce**
These commands show two certificate signing requests, one that works and one that fails.
```
# Input the Vault root token
read -p "Vault root token: " -s VAULT_TOKEN
export VAULT_TOKEN
# Identify the Vault server
export VAULT_ADDR="http://127.0.0.1:8200"
# Set a SAN
OTHER_NAME="1.3.6.1.2.1.27.1.1.2;UTF8:bar"
# Generate a CSR
CSR="$(openssl req -new -newkey rsa:2048 -nodes -config <(<<END cat
[req]
prompt = no
distinguished_name = req_dn
req_extensions = req_ext
[req_dn]
commonName = "foo"
[req_ext]
subjectAltName = @san
[san]
otherName = "$OTHER_NAME"
END
) -keyout "foo.key")"
# Sign the app certificate, repeating the otherName in the JSON payload (which will work)
curl -s --header "X-Vault-Token: $VAULT_TOKEN" --data "$(jq -n --arg CSR "$CSR" --arg OTHER_NAME "$OTHER_NAME" '{ "csr": $CSR, "other_sans": $OTHER_NAME }')" "$VAULT_ADDR/v1/pki/sign/test-role" | jq -r '.data.certificate' | openssl x509 -text | tee "foo-1.crt"
# Sign the app certificate, only specifying the otherName in the CSR (which will create an incorrect certificate with no otherName)
curl -s --header "X-Vault-Token: $VAULT_TOKEN" --data "$(jq -n --arg CSR "$CSR" '{ "csr": $CSR }')" "$VAULT_ADDR/v1/pki/sign/test-role" | jq -r '.data.certificate' | openssl x509 -text | tee "foo-2.crt"
```

**Expected behavior**
Both certificates should contain an `otherName` field.

**Environment:**
* Vault Server Version: 0.11.5
* Vault CLI Version: v0.11.5
* Server Operating System/Architecture: Ubuntu 16.04.5 LTS

### PR DESCRIPTION
**Pull request by ncabatoff**
If a CSR contains a SAN of type otherName, encoded in UTF-8, and the signing role specifies use_csr_sans, the otherName SAN will be included in the signed cert's SAN extension.  Fixes #5989.

I'm still mulling over what to do about tests and whether to trim down the code borrowed from certificate-transparency-go, but the the core logic is ready for review.

### CODE REVIEW

File: builtin/logical/pki/path_issue_sign.go [hunk 1]
```diff
@@ -224,6 +224,8 @@ func (b *backend) pathIssueSignCert(ctx context.Context, req *logical.Request, d
 			return logical.ErrorResponse(err.Error()), nil
 		case errutil.InternalError:
 			return nil, err
+		default:
```
Comment (jefferai) [inline builtin/logical/pki/path_issue_sign.go]: No panics please.

File: builtin/logical/pki/cert_util.go [hunk 2]
```diff
@@ -733,6 +734,173 @@ func signCert(b *backend,
...
+	tagDNSName       = 2
+	tagX400Address   = 3
+	tagDirectoryName = 4
+	tagEDIPartyName  = 5
+	tagURI           = 6
+	tagIPAddress     = 7
+	tagRegisteredID  = 8
+)
+
+// parseGeneralName lifted from https://github.com/google/certificate-transparency-go
```
Comment (jefferai) [inline builtin/logical/pki/cert_util.go]: I don't know whether we can simply take code from an apache product, but my guess is not. I'm also guessing this is overkill for the ask considering that Go's libs already fill in the majority of these and some of these we have no reason to actually support.

File: builtin/logical/pki/path_issue_sign.go (hunk 1 above)
Comment (jefferai) [reply builtin/logical/pki/path_issue_sign.go]: In that case, please fix it to something else :-)

**Comment by jefferai** [comment]
The best way forward I think would be to ask the certtrans project if they would be amenable to making the code public in a lib that we could use. That way we don't duplicate functionality here and can take advantage of the other name types.

**Comment by ncabatoff** [comment]
Despite their overall project's Apache2 license, their embedded x509 fork retains the BSD-style license headers in its files.  Similarly, I found another project that also fulfils our needs: github.com/zmap/zcrypto, which is explicit in its LICENSE file that:

> ZCrypto contains a fork of
several packages from Golang standard library, as well as code from the
BoringSSL test runner. Files that were created by Google, and new files in
forks of packages maintained by Google have a Google copyright and fall under
the ISC license.

Since I've seen code in Vault already that incorporates go stdlib code (to use private bits), I'm guessing that it's not actually a problem to bundle such files.

If you'd really rather avoid the duplication either way, I'm okay with that.  I think ZCrypto is a better bet for us: it's already in the form of a library we could incorporate, and they already provide a function that does exactly what we need (parseGeneralNames), it just needs to be made public.

**Comment by ncabatoff** [comment]
While working on this PR I discovered a bug. The docs say

> allowed_other_sans (string: "") – Defines allowed custom OID/UTF8-string SANs. This field supports globbing. The format is the same as OpenSSL: <oid>;<type>:<value> where the only current valid type is UTF8 (or UTF-8). This can be a comma-delimited list or a JSON string slice. All values, including globbing values, must use the correct syntax, with the exception being a single * which allows any OID and any value (but type must still be UTF8).

In practice however, the only form of globbing that works with allowed_other_sans is "someoid;utf-8:*".  You can't put "*;utf-8:somevalue" or "*;utf-8;*", nor can you use a single star as in "*".  Using a single star results in `error parsing allowed_other_sans: expected a semicolon in other SAN "*"`. Using a star for oid does not produce an error, but it does not have the intended effect of allowing any oid.  This can be demonstrated by modifying the variable named `allowAll` in my new tests.

File: builtin/logical/pki/cert_util.go (hunk 2 above)
Comment (ncabatoff) [reply builtin/logical/pki/cert_util.go]: Went a different route, no longer using code from that project.

File: builtin/logical/pki/path_issue_sign.go (hunk 1 above)
Comment (ncabatoff) [reply builtin/logical/pki/path_issue_sign.go]: I don't understand what you're looking for.  Should I wrap an unexpected error type with an InternalError?

**Comment by jefferai** [comment]
> I think ZCrypto is a better bet for us: it's already in the form of a library we could incorporate, and they already provide a function that does exactly what we need (parseGeneralNames), it just needs to be made public.

Sounds good, can you ask them to make it public?

**Comment by ncabatoff** [comment]
> > I think ZCrypto is a better bet for us: it's already in the form of a library we could incorporate, and they already provide a function that does exactly what we need (parseGeneralNames), it just needs to be made public.
> 
> Sounds good, can you ask them to make it public?

Ah, sorry, I should've made another comment explaining what I did with the code since that comment.  Like I said above, I've seen code in Vault already that incorporates go stdlib code (to use private bits), so I'm guessing that kosher.  So I borrowed forEachSAN from Go's crypto/x509/x509.go and wrote my own getOtherSANsFromX509Extensions based on it.

I got scared off zcrypto once I saw the big DANGER: EXPERIMENTAL warning in its readme, and the text "ZCrypto is a research library, designed to be used for data collection and analysis, as well as experimenting and prototyping. It should not be used to provide security for production systems."

**Comment by catsby** [comment]
Hey @ncabatoff and @jefferai - It sounds like this is ready to merge? I updated this branch from master and the tests pass locally, at least. Let me know if this is ready to merge then.

```
[vault][vlt191-use-csr-sans-othernames](4)$ gotestsum --format short-verbose -- ./builtin/logical/pki

PASS builtin/logical/pki.TestPKI_RequireCN (4.72s)
PASS builtin/logical/pki.TestBackend_CSRValues (0.92s)
PASS builtin/logical/pki.TestBackend_URLsCRUD (0.49s)
PASS builtin/logical/pki.TestBackend_Roles/RSA (8.49s)
PASS builtin/logical/pki.TestBackend_Roles/RSACSR (4.57s)
PASS builtin/logical/pki.TestBackend_Roles/EC (4.62s)
PASS builtin/logical/pki.TestBackend_Roles/ECCSR (2.42s)
PASS builtin/logical/pki.TestBackend_Roles (20.09s)
PASS builtin/logical/pki.TestBackend_PathFetchCertList (2.01s)
PASS builtin/logical/pki.TestBackend_SignVerbatim (0.21s)
PASS builtin/logical/pki.TestBackend_Root_Idempotency (4.70s)
PASS builtin/logical/pki.TestBackend_SignIntermediate_AllowedPastCA (4.20s)
PASS builtin/logical/pki.TestBackend_SignSelfIssued (0.23s)
PASS builtin/logical/pki.TestBackend_OID_SANs (4.71s)
PASS builtin/logical/pki.TestBackend_AllowedSerialNumbers (4.77s)
PASS builtin/logical/pki.TestBackend_URI_SANs (4.21s)
PASS builtin/logical/pki.TestBackend_CA_Steps/teststeps/rsa (25.24s)
PASS builtin/logical/pki.TestBackend_CA_Steps/teststeps/ec (25.29s)
PASS builtin/logical/pki.TestBackend_CA_Steps/teststeps (0.00s)
PASS builtin/logical/pki.TestBackend_CA_Steps (29.35s)
PASS builtin/logical/pki.TestPki_FetchCertBySerial (0.00s)
PASS builtin/logical/pki.TestPki_MultipleOUs (0.00s)
PASS builtin/logical/pki.TestBackend_CRL_EnableDisable (5.21s)
PASS builtin/logical/pki.TestPki_RoleGenerateLease (0.00s)
PASS builtin/logical/pki.TestPki_RoleKeyUsage (0.00s)
PASS builtin/logical/pki.TestPki_RoleOUOrganizationUpgrade (0.00s)
PASS builtin/logical/pki.TestPki_RoleAllowedDomains (0.00s)
PASS builtin/logical/pki.TestPki_RoleAllowedURISANs (0.00s)
PASS builtin/logical/pki.TestPki_RolePkixFields (0.00s)
PASS builtin/logical/pki.TestPki_RoleNoStore (0.57s)
PASS builtin/logical/pki.TestPki_CertsLease (0.65s)
PASS builtin/logical/pki

DONE 31 tests in 91.571s
```

**Comment by ncabatoff** [comment]
It's ready to merge as far as I'm concerned.  I wasn't sure if I should go ahead and do so or wait for final review post-conflicts.

**Comment by malnick** [comment]
@ncabatoff can you check the `test-go` failure? Looks like a SANs failure, but given @jefferai comment above, lets merge once we get that fixed up. Thanks.

**Comment by ncabatoff** [comment]
> @ncabatoff can you check the test-go failure? Looks like a SANs failure, but given @jefferai comment above, lets merge once we get that fixed up. Thanks.

I'm very confused by that error.  It wasn't present in the test run from my last merge, and I don't see any change since that could have introduced it.  It's not happening locally for myself or @catsby.  Maybe it's a glitch in our circleci build output caching?  Trying to rerun the workflow using their UI, so far unsuccessfully.

File: sdk/go.mod [hunk 3]
```diff
@@ -30,6 +30,7 @@ require (
 	golang.org/x/net v0.0.0-20190404232315-eb5bcb51f2a3 // indirect
 	golang.org/x/sys v0.0.0-20190403152447-81d4e9dc473e
 	golang.org/x/text v0.3.1-0.20181227161524-e6919f6577db // indirect
+	google.golang.org/appengine v1.4.0 // indirect
```
Comment (michelvocks) [inline sdk/go.mod]: I saw this change for a while and it feels like everyone is afraid to actually commit it. You wanna do it now? 😄

Comment (ncabatoff) [reply sdk/go.mod]: Sure, I hate having phantom go.mod entries that pop up after I pull.

**Pull request merged**